### PR TITLE
Implement utf8mb4_0900_ai_ci collation

### LINKED ISSUE #7958: Support utf8mb4_0900_ai_ci collation
## Enhancement

TiDB and TiKV support `utf8mb4_0900_ai_ci` collation. TiFlash should also support it.

### PR DESCRIPTION
**Pull request by YangKeao**
### What problem does this PR solve?

Issue Number: close #7958

Problem Summary:

Add the support of `utf8mb4_0900_ai_ci` collation to the TiFlash.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

```release-note
Support the `utf8mb4_0900_ai_ci` collation.
```

### CODE REVIEW

File: dbms/src/Storages/Transaction/tests/gtest_tidb_collator.cpp [hunk 1]
```diff
@@ -92,158 +96,170 @@ const typename CollatorCases::SortKeyCase CollatorCases::sk_cases[] = {
       PREVENT_TRUNC("\x00\x46\x00\x4f\x00\x4f\x00\x20\x00\xa9\x00\x20\x00\x42\x00\x41\x00\x52\x00\x20\xff\xfd\x00\x20"
                     "\x00\x42\x00\x41"
                     "\x00\x5a\x00\x20\x26\x03\x00\x20\x00\x51\x00\x55\x00\x58"),
+<<<<<<< HEAD
```
Comment (windtalker) [inline dbms/src/Storages/Transaction/tests/gtest_tidb_collator.cpp]: Need resolve it

File: dbms/src/Storages/Transaction/Collator.cpp [hunk 2]
```diff
@@ -337,19 +338,59 @@ const std::array<long_weight, 23> weight_lut_long
...
+       long_weight{0x02091C8F1DB91C3F, 0x00001E331C7A1E71},
+       long_weight{0x1E3302091D321D18, 0x000000001E711CAA},
+       long_weight{0x1E711E711DDD1D77, 0x1E711E711CAA1D77},
+       // for default use
+       long_weight{0x0, 0x0}};
 } // namespace UnicodeCI
 
-class UnicodeCICollator final : public ITiDBCollator
+template <typename T, bool padding>
+class UCACICollator final : public ITiDBCollator
```
Comment (windtalker) [inline dbms/src/Storages/Transaction/Collator.cpp:374]: What is `UCA` mean?

File: dbms/src/Storages/Transaction/Collator.cpp [hunk 3]
```diff
@@ -402,7 +443,15 @@ class UnicodeCICollator final : public ITiDBCollator
...
-        auto v = rtrim(s, length);
+        std::string_view v;
+        if (padding)
+        {
+            v = rtrim(s, length);
+        }
+        else
+        {
+            v = std::string_view(s, length);
+        }
```
Comment (windtalker) [inline dbms/src/Storages/Transaction/Collator.cpp:454]: Better to have use a function todo this, so the we don't need to copy code here

File: dbms/src/Storages/Transaction/Collator.cpp (hunk 2 above)
Comment (YangKeao) [reply dbms/src/Storages/Transaction/Collator.cpp:374]: [Unicode Collation Algorithm](https://unicode.org/reports/tr10/), the standard of the unicode collation.

File: dbms/src/Storages/Transaction/Collator.cpp (hunk 3 above)
Comment (YangKeao) [reply dbms/src/Storages/Transaction/Collator.cpp]: Done. PTAL @windtalker

File: dbms/src/Storages/Transaction/Collator.cpp [hunk 4]
```diff
@@ -444,15 +475,56 @@ class UnicodeCICollator final : public ITiDBCollator
...
+            {
+                first = second;
+                second = 0;
+            }
+        }
+    }
+
+    std::string_view preprocess(const char * s, size_t length) const
+    {
+        if (padding) {
```
Comment (windtalker) [inline dbms/src/Storages/Transaction/Collator.cpp:506]: use `constexpr` here?

Comment (YangKeao) [reply dbms/src/Storages/Transaction/Collator.cpp]: Done.

**Review by windtalker — APPROVED** [review]
LGTM

**Review by yibin87 — APPROVED** [review]
LGTM

**Pull request merged**